repo2docker/contentproviders/hydroshare.py

```diff
@@ -0,0 +1,110 @@
+import zipfile
+import os
+import shutil
+import time
+import json
+from datetime import datetime, timezone, timedelta
+
+from urllib.request import urlopen, Request, urlretrieve
+from urllib.error import HTTPError
+
+from .doi import DoiProvider
+from .base import ContentProviderException
+from ..utils import normalize_doi, is_doi
+
+
+class Hydroshare(DoiProvider):
+    """Provide contents of a Hydroshare resource."""
+
+    def detect(self, doi, ref=None, extra_args=None):
+        """Trigger this provider for things that resolve to a Hydroshare resource"""
+        hosts = [
+            {
+                "hostname": [
+                    "https://www.hydroshare.org/resource/",
+                    "http://www.hydroshare.org/resource/",
+                ],
+                "django_irods": "https://www.hydroshare.org/django_irods/download/bags/",
+                "version": "https://www.hydroshare.org/hsapi/resource/{}/scimeta/elements",
+            }
+        ]
+
+        def fetch_version(resource_id, host):
+            """Fetch resource modified date and convert to epoch"""
+            json_response = json.loads(
+                self.urlopen(host["version"].format(self.resource_id)).read()
+            )
+            date = next(
+                item for item in json_response["dates"] if item["type"] == "modified"
+            )["start_date"]
+            # Hydroshare timestamp always returns the same timezone, so strip it
+            date = date.split(".")[0]
+            parsed_date = datetime.strptime(date, "%Y-%m-%dT%H:%M:%S")
+            epoch = parsed_date.replace(tzinfo=timezone(timedelta(0))).timestamp()
+            # truncate the timestamp
+            return str(int(epoch))
+
+        url = self.doi2url(doi)
+
+        for host in hosts:
+            if any([url.startswith(s) for s in host["hostname"]]):
+                self.resource_id = url.strip("/").rsplit("/", maxsplit=1)[1]
+                self.version = fetch_version(self.resource_id, host)
+                return {
+                    "resource": self.resource_id,
+                    "host": host,
+                    "version": self.version,
+                }
+
+    def _urlretrieve(self, bag_url):
+        return urlretrieve(bag_url)
+
+    def fetch(self, spec, output_dir, yield_output=False, timeout=120):
+        """Fetch and unpack a Hydroshare resource"""
+        resource_id = spec["resource"]
+        host = spec["host"]
+
+        yield "Fetching HydroShare Resource {}.\n".format(resource_id)
+
+        bag_url = "{}{}".format(host["django_irods"], resource_id)
```

Comment:
From a user perspective, I would like to see the full URL also printed so I know that the download has started, maybe `yield "Downloading {}.}n".format(bag_url)`.

The DOI from the tests actually returns a 2.7 GB large repo... I know it is not actually downloaded during the tests... but maybe we can (also) have a smaller one there for tests?

Reply:
@nuest - Thanks for taking a look.  Could you clarify what you'd like to test with the smaller resource?  Do you just want me to add a comment pointing to a smaller resource to test locally with?  I'm not sure what you're looking for.

Reply:
Yes, something smaller for manual testing. 

_BUT maybe also for a test_: I'm not sure what the state of the other DOI-based providers is, but it could be worth having at least one "integration test" that actually tries to create a Binder from a DOI. @betatim What's your take? Should we open a new issue on that?

In any case, that should not hinder this PR. I might be the only person you looks into the source code for a example DOI anyway.

Reply:
Here's a much smaller DOI.  https://doi.org/10.4211/hs.142c59757ed54de1816777828c9716e7

Reply:
For "integration" tests we usually make a copy of https://github.com/binder-examples/requirements and put that into what ever "storage" we want to test. Generally we want to test as much as possible with unit tests though or tests that fake stuff because the repo2docker tests as a whole take long enough to execute that glaciers melt while we do that.

So I'd say we don't need to check that r2d will successfully build a container from a known good content (like binder-examples/requirements). What we could test "integration test" style is actually downloading a small file from hydroshare to catch when our fake download tests are outdated.